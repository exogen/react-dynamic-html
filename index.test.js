import React from "react";
import { render } from "react-testing-library";
let Template;

function defineTests() {
  afterEach(() => {
    Template.resetTemplateId();
  });

  it("renders the given string into a div", () => {
    const { container } = render(<Template string="Hi there!" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders the given string into the supplied DOM element type", () => {
    const { container } = render(<Template string="Hi there!" as="span" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("passes props to the rendered element", () => {
    const { container } = render(
      <Template className="StyledTemplate" string="Hi there!" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders placeholders", () => {
    const { container } = render(
      <Template
        string="<h1>Chapter 1: {title}</h1>"
        values={{
          title: "A Good Test"
        }}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders React placeholders", () => {
    const { container } = render(
      <Template
        string="<h1>Chapter 1: {title}</h1>"
        values={{
          title: (
            <React.Fragment>
              A <em>Better</em> Test
            </React.Fragment>
          )
        }}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders value updates", () => {
    const { container, rerender } = render(
      <Template
        string="Hey! {title}"
        values={{
          title: <h1>A Good Test</h1>
        }}
      />
    );
    rerender(
      <Template string="Hey! {title}" values={{ title: <h2>A Bad Test</h2> }} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders template updates", () => {
    const { container, rerender } = render(
      <Template
        string="Hey! {title}"
        values={{
          title: <h1>A Good Test</h1>
        }}
      />
    );
    rerender(
      <Template
        string="Hey! {title} …and again… {title}"
        values={{
          title: <h1>A Good Test</h1>
        }}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders numbers, undefined, null, and objects", () => {
    const { container } = render(
      <Template
        string="Types: {number}, {undefined}, {null}, {object}"
        values={{
          number: 0,
          null: null,
          object: [1, 2, 3]
        }}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders with the correct context", () => {
    const Context = React.createContext("FAIL");
    const consumer = (
      <Context.Consumer>
        {value => <h1 data-testid="consumer">{value}</h1>}
      </Context.Consumer>
    );
    const { getByTestId } = render(
      <Context.Provider value="PASS">
        <Template string="<div>{consumer}</div>" values={{ consumer }} />
      </Context.Provider>
    );
    expect(getByTestId("consumer")).toHaveTextContent("PASS");
  });

  it("can render values into a different host tag", () => {
    const { container } = render(
      <Template
        string="Hey! {title}"
        values={{ title: <h1>A Good Test</h1> }}
        valueTags={{ title: "header" }}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("can supply a default host tag", () => {
    const { container } = render(
      <Template
        string="Hey! {title}"
        values={{ title: <h1>A Good Test</h1> }}
        defaultValueTag="section"
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("can render templates inside templates", () => {
    const { container } = render(
      <Template
        string="A: {test}"
        values={{
          test: (
            <Template
              string="B: {test}"
              values={{
                test: (
                  <Template
                    string="C: {test}"
                    values={{ test: <span>OK!</span> }}
                  />
                )
              }}
            />
          )
        }}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
}

describe("Template", () => {
  describe("server", () => {
    beforeAll(() => {
      jest.resetModules();
      process.browser = false;
      Template = require("./index").default;
    });

    defineTests();
  });

  describe("browser", () => {
    beforeAll(() => {
      jest.resetModules();
      process.browser = true;
      Template = require("./index").default;
    });

    defineTests();
  });
});
