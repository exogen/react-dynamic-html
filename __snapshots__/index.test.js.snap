// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template class client-side can render templates inside templates 1`] = `
<div>
  A: 
  <span
    data-template-id="1"
    data-template-key="test:1"
  >
    <div>
      B: 
      <span
        data-template-id="2"
        data-template-key="test:1"
      >
        <div>
          C: 
          <span
            data-template-id="3"
            data-template-key="test:1"
          >
            <span>
              OK!
            </span>
          </span>
        </div>
      </span>
    </div>
  </span>
</div>
`;

exports[`Template class client-side can render values into a different host tag 1`] = `
<div>
  Hey! 
  <header
    data-template-id="1"
    data-template-key="title:1"
  >
    <h1>
      A Good Test
    </h1>
  </header>
</div>
`;

exports[`Template class client-side can supply a default host tag 1`] = `
<div>
  Hey! 
  <section
    data-template-id="1"
    data-template-key="title:1"
  >
    <h1>
      A Good Test
    </h1>
  </section>
</div>
`;

exports[`Template class client-side does not unmount moved portals 1`] = `
<div>
  <div>
    <span
      data-template-id="1"
      data-template-key="button:1"
    >
      <button
        data-testid="button"
      >
        2
      </button>
    </span>
  </div>
</div>
`;

exports[`Template class client-side passes props to the rendered element 1`] = `
<div
  class="StyledTemplate"
>
  Hi there!
</div>
`;

exports[`Template class client-side renders React placeholders 1`] = `
<div>
  <h1>
    Chapter 1: 
    <span
      data-template-id="1"
      data-template-key="title:1"
    >
      A 
      <em>
        Better
      </em>
       Test
    </span>
  </h1>
</div>
`;

exports[`Template class client-side renders numbers, undefined, null, arrays, and objects 1`] = `
<div>
  Types: 0, , , 
  <span
    data-template-id="1"
    data-template-key="array:1"
  >
    1
    2
    3
  </span>
  , [object Hello]
</div>
`;

exports[`Template class client-side renders placeholders 1`] = `
<div>
  <h1>
    Chapter 1: A Good Test
  </h1>
</div>
`;

exports[`Template class client-side renders template updates 1`] = `
<div>
  Hey! 
  <span
    data-template-id="1"
    data-template-key="title:1"
  >
    <h1>
      A Good Test
    </h1>
  </span>
   …and again… 
  <span
    data-template-id="1"
    data-template-key="title:2"
  >
    <h1>
      A Good Test
    </h1>
  </span>
</div>
`;

exports[`Template class client-side renders the given string into a div 1`] = `
<div>
  Hi there!
</div>
`;

exports[`Template class client-side renders the given string into the supplied DOM element type 1`] = `
<span>
  Hi there!
</span>
`;

exports[`Template class client-side renders value updates 1`] = `
<div>
  Hey! 
  <span
    data-template-id="1"
    data-template-key="title:1"
  >
    <h2>
      A Bad Test
    </h2>
  </span>
</div>
`;

exports[`Template class client-side supports string value patterns 1`] = `
<div>
  Hey! 
  <span
    data-template-id="1"
    data-template-key="title:1"
  >
    <h1>
      It worked!
    </h1>
  </span>
</div>
`;

exports[`Template class server-side can render templates inside templates 1`] = `
<div>
  A: 
  <span>
    <div>
      B: 
      <span>
        <div>
          C: 
          <span>
            <span>
              OK!
            </span>
          </span>
        </div>
      </span>
    </div>
  </span>
</div>
`;

exports[`Template class server-side can render values into a different host tag 1`] = `
<div>
  Hey! 
  <header>
    <h1>
      A Good Test
    </h1>
  </header>
</div>
`;

exports[`Template class server-side can supply a default host tag 1`] = `
<div>
  Hey! 
  <section>
    <h1>
      A Good Test
    </h1>
  </section>
</div>
`;

exports[`Template class server-side passes props to the rendered element 1`] = `
<div
  class="StyledTemplate"
>
  Hi there!
</div>
`;

exports[`Template class server-side renders React placeholders 1`] = `
<div>
  <h1>
    Chapter 1: 
    <span>
      A 
      <em>
        Better
      </em>
       Test
    </span>
  </h1>
</div>
`;

exports[`Template class server-side renders numbers, undefined, null, arrays, and objects 1`] = `
<div>
  Types: 0, , , 
  <span>
    1
    2
    3
  </span>
  , [object Hello]
</div>
`;

exports[`Template class server-side renders placeholders 1`] = `
<div>
  <h1>
    Chapter 1: A Good Test
  </h1>
</div>
`;

exports[`Template class server-side renders the given string into a div 1`] = `
<div>
  Hi there!
</div>
`;

exports[`Template class server-side renders the given string into the supplied DOM element type 1`] = `
<span>
  Hi there!
</span>
`;

exports[`Template class server-side supports string value patterns 1`] = `
<div>
  Hey! 
  <span>
    <h1>
      It worked!
    </h1>
  </span>
</div>
`;

exports[`Template using hooks client-side can render templates inside templates 1`] = `
<div>
  A: 
  <span
    data-template-id="1"
    data-template-key="test:1"
  >
    <div>
      B: 
      <span
        data-template-id="2"
        data-template-key="test:1"
      >
        <div>
          C: 
          <span
            data-template-id="3"
            data-template-key="test:1"
          >
            <span>
              OK!
            </span>
          </span>
        </div>
      </span>
    </div>
  </span>
</div>
`;

exports[`Template using hooks client-side can render values into a different host tag 1`] = `
<div>
  Hey! 
  <header
    data-template-id="1"
    data-template-key="title:1"
  >
    <h1>
      A Good Test
    </h1>
  </header>
</div>
`;

exports[`Template using hooks client-side can supply a default host tag 1`] = `
<div>
  Hey! 
  <section
    data-template-id="1"
    data-template-key="title:1"
  >
    <h1>
      A Good Test
    </h1>
  </section>
</div>
`;

exports[`Template using hooks client-side does not unmount moved portals 1`] = `
<div>
  <div>
    <span
      data-template-id="1"
      data-template-key="button:1"
    >
      <button
        data-testid="button"
      >
        2
      </button>
    </span>
  </div>
</div>
`;

exports[`Template using hooks client-side passes props to the rendered element 1`] = `
<div
  class="StyledTemplate"
>
  Hi there!
</div>
`;

exports[`Template using hooks client-side renders React placeholders 1`] = `
<div>
  <h1>
    Chapter 1: 
    <span
      data-template-id="1"
      data-template-key="title:1"
    >
      A 
      <em>
        Better
      </em>
       Test
    </span>
  </h1>
</div>
`;

exports[`Template using hooks client-side renders numbers, undefined, null, arrays, and objects 1`] = `
<div>
  Types: 0, , , 
  <span
    data-template-id="1"
    data-template-key="array:1"
  >
    1
    2
    3
  </span>
  , [object Hello]
</div>
`;

exports[`Template using hooks client-side renders placeholders 1`] = `
<div>
  <h1>
    Chapter 1: A Good Test
  </h1>
</div>
`;

exports[`Template using hooks client-side renders template updates 1`] = `
<div>
  Hey! 
  <span
    data-template-id="1"
    data-template-key="title:1"
  >
    <h1>
      A Good Test
    </h1>
  </span>
   …and again… 
  <span
    data-template-id="1"
    data-template-key="title:2"
  >
    <h1>
      A Good Test
    </h1>
  </span>
</div>
`;

exports[`Template using hooks client-side renders the given string into a div 1`] = `
<div>
  Hi there!
</div>
`;

exports[`Template using hooks client-side renders the given string into the supplied DOM element type 1`] = `
<span>
  Hi there!
</span>
`;

exports[`Template using hooks client-side renders value updates 1`] = `
<div>
  Hey! 
  <span
    data-template-id="1"
    data-template-key="title:1"
  >
    <h2>
      A Bad Test
    </h2>
  </span>
</div>
`;

exports[`Template using hooks client-side supports string value patterns 1`] = `
<div>
  Hey! 
  <span
    data-template-id="1"
    data-template-key="title:1"
  >
    <h1>
      It worked!
    </h1>
  </span>
</div>
`;

exports[`Template using hooks server-side can render templates inside templates 1`] = `
<div>
  A: 
  <span>
    <div>
      B: 
      <span>
        <div>
          C: 
          <span>
            <span>
              OK!
            </span>
          </span>
        </div>
      </span>
    </div>
  </span>
</div>
`;

exports[`Template using hooks server-side can render values into a different host tag 1`] = `
<div>
  Hey! 
  <header>
    <h1>
      A Good Test
    </h1>
  </header>
</div>
`;

exports[`Template using hooks server-side can supply a default host tag 1`] = `
<div>
  Hey! 
  <section>
    <h1>
      A Good Test
    </h1>
  </section>
</div>
`;

exports[`Template using hooks server-side passes props to the rendered element 1`] = `
<div
  class="StyledTemplate"
>
  Hi there!
</div>
`;

exports[`Template using hooks server-side renders React placeholders 1`] = `
<div>
  <h1>
    Chapter 1: 
    <span>
      A 
      <em>
        Better
      </em>
       Test
    </span>
  </h1>
</div>
`;

exports[`Template using hooks server-side renders numbers, undefined, null, arrays, and objects 1`] = `
<div>
  Types: 0, , , 
  <span>
    1
    2
    3
  </span>
  , [object Hello]
</div>
`;

exports[`Template using hooks server-side renders placeholders 1`] = `
<div>
  <h1>
    Chapter 1: A Good Test
  </h1>
</div>
`;

exports[`Template using hooks server-side renders the given string into a div 1`] = `
<div>
  Hi there!
</div>
`;

exports[`Template using hooks server-side renders the given string into the supplied DOM element type 1`] = `
<span>
  Hi there!
</span>
`;

exports[`Template using hooks server-side supports string value patterns 1`] = `
<div>
  Hey! 
  <span>
    <h1>
      It worked!
    </h1>
  </span>
</div>
`;
